### PR TITLE
xds: WRR in xDS

### PR DESCRIPTION
**Pull request by zasweq**
This PR does a few things:

* Prepare JSON safely in xDS LB policy registry (prepare a struct, then call json.Marshal on that struct to sanitize)
* Add WRR Configuration to this xDS LB policy registry and a sanity e2e test case for WRR in xDS)

Notes:

* The config preparation through structs is determined by which ones prepare with untrusted inputs. Other preparations, such as name: balancer name, value: json, don't need to go through a struct since json here is already assumed to be sanitized.
* I could not add a unit test for converter logic in converter_test or unmarshal_cds, as it tests the JSON by unmarshaling using internalserviceconfig.BalancerConfig and comparing against an exported config want (put in e2e_test to not take dependency cycle in main code). However, the wrr config is not exported.

RELEASE NOTES: N/A

### CODE REVIEW

File: test/xds/xds_client_custom_lb_test.go [hunk 1]
```diff
@@ -180,6 +184,66 @@ func (s) TestWrrLocality(t *testing.T) {
...
+			wrrLocalityConfiguration: wrrLocality(&v3clientsideweightedroundrobinpb.ClientSideWeightedRoundRobin{
+				EnableOobLoadReport: &wrapperspb.BoolValue{
+					Value: false,
+				},
+				// BlackoutPeriod long enough to cause load report weights to
+				// trigger in the scope of test case, but no load reports
+				// configured anyway.
+				BlackoutPeriod:          &durationpb.Duration{Seconds: 10},
+				WeightExpirationPeriod:  &durationpb.Duration{Seconds: 10},
+				WeightUpdatePeriod:      &durationpb.Duration{Seconds: 1},
```
Comment (dfawley) [inline test/xds/xds_client_custom_lb_test.go:205]: Can you use `durationpb.New(10*time.Seconds)` instead?

File: test/xds/xds_client_custom_lb_test.go [hunk 2]
```diff
@@ -180,6 +184,66 @@ func (s) TestWrrLocality(t *testing.T) {
...
+				ErrorUtilizationPenalty: &wrapperspb.FloatValue{Value: 1},
+			}),
+			addressDistributionWant: []resolver.Address{
```
Comment (dfawley) [inline test/xds/xds_client_custom_lb_test.go:208]: This is pretty awkward to write and to read.  Can you make this a struct that has the address & count instead?

File: xds/internal/xdsclient/xdslbregistry/converter.go [hunk 3]
```diff
@@ -172,6 +204,65 @@ func convertCustomPolicy(typeURL string, s *structpb.Struct) (json.RawMessage, b
...
+		WeightExpirationPeriod:  internalserviceconfig.Duration(weightExpirationPeriod),
+		WeightUpdatePeriod:      internalserviceconfig.Duration(weightUpdatePeriod),
+		ErrorUtilizationPenalty: float64(errorUtilizationPenalty),
+	}
+
+	lbCfgJSON, err := json.Marshal(wrrLBConfig)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling JSON for type %T: %v", wrrLBConfig, err)
+	}
+	return makeBalancerConfigJSON("weighted_round_robin_experimental", lbCfgJSON), nil
```
Comment (dfawley) [inline xds/internal/xdsclient/xdslbregistry/converter.go:263]: This can use `weightedroundrobin.Name` for sure.

File: xds/internal/xdsclient/xdslbregistry/converter.go [hunk 4]
```diff
@@ -172,6 +204,65 @@ func convertCustomPolicy(typeURL string, s *structpb.Struct) (json.RawMessage, b
...
+
+func convertClientSideWRR(cfg *v3clientsideweightedroundrobinpb.ClientSideWeightedRoundRobin) (json.RawMessage, error) {
+	const (
+		defaultEnableOOBLoadReport     = false
+		defaultOOBReportingPeriod      = 10 * time.Second
+		defaultBlackoutPeriod          = 10 * time.Second
+		defaultWeightExpirationPeriod  = 3 * time.Minute
+		defaultWeightUpdatePeriod      = time.Second
+		defaultErrorUtilizationPenalty = float32(1.0)
+	)
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/xdslbregistry/converter.go:224]: I'll make it zero value by just declaring var x, I won't do x := 0

File: test/xds/xds_client_custom_lb_test.go (hunk 1 above)
Comment (zasweq) [reply test/xds/xds_client_custom_lb_test.go:205]: Done for all 3.

File: xds/internal/xdsclient/xdslbregistry/converter.go [hunk 5]
```diff
@@ -132,16 +146,34 @@ func convertRingHash(cfg *v3ringhashpb.RingHash) (json.RawMessage, error) {
...
+	rhCfg := &ringHashLBConfig{
+		MinRingSize: minSize,
+		MaxRingSize: maxSize,
+	}
+
+	rhCfgJSON, err := json.Marshal(rhCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling JSON for type %T: %v", rhCfg, err)
+	}
+	return makeBalancerConfigJSON("ring_hash_experimental", rhCfgJSON), nil
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/xdslbregistry/converter.go:158]: Switched.

File: xds/internal/xdsclient/xdslbregistry/converter.go [hunk 6]
```diff
@@ -132,16 +146,34 @@ func convertRingHash(cfg *v3ringhashpb.RingHash) (json.RawMessage, error) {
...
 	}
-	lbCfgJSON := []byte(fmt.Sprintf(`{"childPolicy": %s}`, epJSON))
+	wrrLCfg := wrrLocalityLBConfig{
+		ChildPolicy: epJSON,
+	}
+	lbCfgJSON, err := json.Marshal(wrrLCfg)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling JSON for type %T: %v", wrrLCfg, err)
+	}
 	return makeBalancerConfigJSON("xds_wrr_locality_experimental", lbCfgJSON), nil
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/xdslbregistry/converter.go:177]: Switched.

File: xds/internal/xdsclient/xdslbregistry/converter.go (hunk 3 above)
Comment (zasweq) [reply xds/internal/xdsclient/xdslbregistry/converter.go:263]: Switched.

File: xds/internal/xdsclient/xdslbregistry/converter.go [hunk 7]
```diff
@@ -86,6 +89,12 @@ func convertToServiceConfig(lbPolicy *v3clusterpb.LoadBalancingPolicy, depth int
 				return nil, fmt.Errorf("failed to unmarshal resource: %v", err)
 			}
 			return convertWrrLocality(wrrlProto, depth)
+		case "type.googleapis.com/envoy.extensions.load_balancing_policies.client_side_weighted_round_robin.v3.ClientSideWeightedRoundRobin":
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdsclient/xdslbregistry/converter.go:92]: Switched to a registry.

File: test/xds/xds_client_custom_lb_test.go (hunk 2 above)
Comment (zasweq) [reply test/xds/xds_client_custom_lb_test.go:208]: Done.

**Comment by dfawley** [comment]
`xds/internal/xdslbregistry/converter.go:167:2: field depth is unused (U1000)`

File: test/xds/xds_client_custom_lb_test.go [hunk 8]
```diff
@@ -124,43 +132,30 @@ func (s) TestWrrLocality(t *testing.T) {
...
+					count: 8,
+				},
+				{
+					addr:  backend4.Address,
+					count: 8,
+				},
+				{
+					addr:  backend5.Address,
+					count: 8,
+				},
```
Comment (dfawley) [inline test/xds/xds_client_custom_lb_test.go:158]: Up to you, but this can be a lot easier to read if it's compressed:

```go
  {addr: backend1.Address, count:6},
  {addr: backend2.Address, count:6},
  {addr: backend3.Address, count:8},
  {addr: backend4.Address, count:8},
...
```

File: xds/internal/xdslbregistry/converter.go [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+	"strings"
+
+	v1xdsudpatypepb "github.com/cncf/xds/go/udpa/type/v1"
+	v3xdsxdstypepb "github.com/cncf/xds/go/xds/type/v3"
+	v3clusterpb "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3clientsideweightedroundrobinpb "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/client_side_weighted_round_robin/v3"
+	v3ringhashpb "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/ring_hash/v3"
+	v3wrrlocalitypb "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/wrr_locality/v3"
+	"github.com/golang/protobuf/proto"
+	structpb "github.com/golang/protobuf/ptypes/struct"
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:37]: Nit: Please move the pbs to the end

File: xds/internal/xdslbregistry/converter.go [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+	m = make(map[string]converter)
+)
+
+// converter converts raw proto bytes into the internal Go JSON representation
+// of the proto passed. Returns the json message, an error, and a bool
+// determining if the caller should continue to the next proto in the policy
+// list.
+type converter interface {
+	convertToServiceConfig([]byte, int) (json.RawMessage, bool, error)
+}
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:58]: This is fine.  If you want to simplify, if we don't see a need for state or more functions, then `type converter func([]byte, int)....` works and avoids the need for all the structs you define.

File: xds/internal/xdslbregistry/converter.go [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+	m[strings.ToLower(protoType)] = c
+}
+
+// getConverter returns the converter registered with the given proto type. If
+// no converter is registered with the name, nil is returned.
+func getConverter(name string) converter {
+	if c, ok := m[strings.ToLower(name)]; ok {
+		return c
+	}
+	return nil
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:85]: Equivalent:

```go
return m[strings.ToLower(name)] // or m[name] in which case delete this function even?
```

File: xds/internal/xdslbregistry/converter.go [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+		// - A52
+		// This includes Least Request as well, since grpc-go does not support
+		// the Least Request Load Balancing Policy.
+		if converter == nil {
+			continue
+		}
+		json, cont, err := converter.convertToServiceConfig(policy.GetTypedExtensionConfig().GetTypedConfig().GetValue(), depth)
+		if cont {
+			continue
+		}
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:120]: Can we instead just use `json == nil` for this?  It's less explicit, but it would streamline the API and the intent seems very clear (plus it's all internal to the package).

File: xds/internal/xdslbregistry/converter.go [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+func (wlc *wrrLocalityConverter) convertToServiceConfig(rawProto []byte, depth int) (json.RawMessage, bool, error) {
+	wrrlProto := &v3wrrlocalitypb.WrrLocality{}
+	if err := proto.Unmarshal(rawProto, wrrlProto); err != nil {
+		return nil, false, fmt.Errorf("failed to unmarshal resource: %v", err)
+	}
+	epJSON, err := convertToServiceConfig(wrrlProto.GetEndpointPickingPolicy(), depth+1)
+	if err != nil {
+		return nil, false, fmt.Errorf("error converting endpoint picking policy: %v for %+v", err, wrrlProto)
+	}
+	wrrLCfg := wrrLocalityLBConfig{
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:183]: `wrrL*B*Cfg`?

File: xds/internal/xdslbregistry/converter.go [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+type weightedRoundRobinConverter struct{}
+
+func (wrrc *weightedRoundRobinConverter) convertToServiceConfig(rawProto []byte, depth int) (json.RawMessage, bool, error) {
+	cswrrProto := &v3clientsideweightedroundrobinpb.ClientSideWeightedRoundRobin{}
+	if err := proto.Unmarshal(rawProto, cswrrProto); err != nil {
+		return nil, false, fmt.Errorf("failed to unmarshal resource: %v", err)
+	}
+	wrrLBConfig := &wrrLBConfig{}
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:201]: Nit: `Cfg`?  1. doesn't shadow, 2. matches pattern established.

File: xds/internal/xdslbregistry/converter.go [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+// converted json, a bool representing whether the caller should continue to the
+// next policy, which is true if the gRPC Balancer registry does not contain
+// that balancer type, and an error which should cause caller to error if error
+// converting.
+func convertCustomPolicy(typeURL string, s *structpb.Struct) (json.RawMessage, bool, error) {
+	// The gRPC policy name will be the "type name" part of the value of the
+	// type_url field in the TypedStruct. We get this by using the part after
+	// the last / character. Can assume a valid type_url from the control plane.
+	urls := strings.Split(typeURL, "/")
+	name := urls[len(urls)-1]
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:261]: Nit, but slightly more efficient as it doesn't require allocating a slice:

```go
pos := strings.LastIndex(typeURL, "/")
name := typeURL[pos+1:]
```

File: xds/internal/xdslbregistry/converter_test.go [hunk 16]
```diff
@@ -349,7 +345,7 @@ func (s) TestConvertToServiceConfigFailure(t *testing.T) {
 		},
 		// TODO: test validity right on the boundary of recursion 16 layers
 		// total.
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter_test.go:347]: Can you not just delete one `wrrLocality` from below and take care of this in 30 seconds?

File: xds/internal/xdslbregistry/converter_test.go [hunk 17]
```diff
@@ -349,7 +345,7 @@ func (s) TestConvertToServiceConfigFailure(t *testing.T) {
...
-		{
+		{ // every test case should work except
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter_test.go:348]: This comment seems unnecessary.  It's called "too much" and wants an error.

File: xds/internal/xdslbregistry/converter_test.go [hunk 18]
```diff
@@ -366,7 +362,7 @@ func (s) TestConvertToServiceConfigFailure(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			_, gotErr := xdslbregistry.ConvertToServiceConfig(test.policy)
+			_, gotErr := ConvertToServiceConfig(test.policy)
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter_test.go:365]: Did we wanna make this test be `package xdslbregistry_test` and keep this?  Seems worth it.

File: xds/internal/xdslbregistry/converter_test.go (hunk 16 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter_test.go:347]: One below is like 20+, but I can add, but it seems like a useless test case.

File: xds/internal/xdslbregistry/converter_test.go (hunk 18 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter_test.go:365]: Why? To test at the API layer? The main reason we had this in separate package initially was to not import LB Configs.

File: test/xds/xds_client_custom_lb_test.go (hunk 8 above)
Comment (zasweq) [reply test/xds/xds_client_custom_lb_test.go:158]: Done.

File: xds/internal/xdslbregistry/converter.go [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.round_robin.v3.RoundRobin", &roundRobinConverter{})
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.wrr_locality.v3.WrrLocality", &wrrLocalityConverter{})
+	register("type.googleapis.com/envoy.extensions.load_balancing_policies.client_side_weighted_round_robin.v3.ClientSideWeightedRoundRobin", &weightedRoundRobinConverter{})
+	register("type.googleapis.com/xds.type.v3.TypedStruct", &v3TypedStructConverter{})
+	register("type.googleapis.com/udpa.type.v1.TypedStruct", &v1TypedStructConverter{})
+}
+
+// register registers the converter to the map keyed on a proto type.
+func register(protoType string, c converter) {
+	m[strings.ToLower(protoType)] = c
```
[the remark this answers is not in the document]
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:76]: No not supposed to be case insensitive. I just figured if the balancer registry did it maybe this should to. Switched to inline.

File: xds/internal/xdslbregistry/converter.go (hunk 11 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:85]: Let's do m[name]

File: xds/internal/xdslbregistry/converter.go (hunk 12 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:120]: Ummmmm I don't think that makes sense here (switching bool to invariant of json, which happens to be able to be nil or not nil). Feels weird at every real error you need to return a populated empty byte slice?

File: xds/internal/xdslbregistry/converter.go (hunk 13 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:183]: "weighted round robin locality config"

File: xds/internal/xdslbregistry/converter.go (hunk 14 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:201]: Yeah, good point on 2. Is shadowing types bad? I've def had a lot of type := type declarations.

File: xds/internal/xdslbregistry/converter.go (hunk 15 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:261]: Done.

File: xds/internal/xdslbregistry/converter.go (hunk 10 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:58]: Switched.

File: xds/internal/xdslbregistry/converter_test.go (hunk 18 above)
Comment (dfawley) [reply xds/internal/xdslbregistry/converter_test.go:365]: > To test at the API layer?

Yes?  It seems we don't need any internals, so it seems better to not be in the same package unless necessary.

File: xds/internal/xdslbregistry/converter_test.go (hunk 16 above)
Comment (dfawley) [reply xds/internal/xdslbregistry/converter_test.go:347]: So delete 5 or whatever.

If it's useless, why have a TODO?

Comment (dfawley) [reply xds/internal/xdslbregistry/converter_test.go]: Also, if it's important to test the boundary condition (to have the same behavior as c and java), then we should test [allowable_limit, allowable_limit+1] not [..., allowable_limit+5].

File: xds/internal/xdslbregistry/converter.go (hunk 12 above)
Comment (dfawley) [reply xds/internal/xdslbregistry/converter.go:122]: No.  It would only skip if `err == nil && json == nil`.

File: xds/internal/xdslbregistry/converter_test.go (hunk 18 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter_test.go:365]: K switched.

File: xds/internal/xdslbregistry/converter_test.go (hunk 16 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter_test.go]: K added test and switched this test to + 1.

File: xds/internal/xdslbregistry/converter.go (hunk 14 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go]: Sg.

File: xds/internal/xdslbregistry/converter.go (hunk 12 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:122]: Sounds good. Switched :).

File: xds/internal/xdslbregistry/converter_test.go (hunk 18 above)
Comment (dfawley) [reply xds/internal/xdslbregistry/converter_test.go]: This can go in the same directory.  Please move it back.

File: xds/internal/xdslbregistry/tests/converter_test.go [hunk 20]
```diff
@@ -16,8 +16,8 @@
  *
  */
 
-// Package tests_test contains test cases for the xDS LB Policy Registry.
-package tests_test
```
Comment (dfawley) [inline xds/internal/xdslbregistry/tests/converter_test.go:20]: The convention is `package xdslbregistry_test` in the same directory as `package xdslbregistry`.

File: xds/internal/xdslbregistry/converter.go (hunk 13 above)
Comment (dfawley) [reply xds/internal/xdslbregistry/converter.go:160]: Ohhhh OK sorry

File: xds/internal/xdslbregistry/converter.go [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+	"google.golang.org/grpc/xds/internal/balancer/ringhash"
+	"google.golang.org/grpc/xds/internal/balancer/wrrlocality"
+
+	v1xdsudpatypepb "github.com/cncf/xds/go/udpa/type/v1"
+	v3xdsxdstypepb "github.com/cncf/xds/go/xds/type/v3"
+	v3clusterpb "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3clientsideweightedroundrobinpb "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/client_side_weighted_round_robin/v3"
+	v3ringhashpb "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/ring_hash/v3"
+	v3wrrlocalitypb "github.com/envoyproxy/go-control-plane/envoy/extensions/load_balancing_policies/wrr_locality/v3"
+	"github.com/golang/protobuf/proto"
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:43]: This belongs above though.

File: xds/internal/xdslbregistry/converter.go [hunk 22]
```diff
@@ -0,0 +1,259 @@
...
+func init() {
+	m = map[string]converter{
+		"type.googleapis.com/envoy.extensions.load_balancing_policies.ring_hash.v3.RingHash":                                            convertRingHashProtoToServiceConfig,
+		"type.googleapis.com/envoy.extensions.load_balancing_policies.round_robin.v3.RoundRobin":                                        convertRoundRobinProtoToServiceConfig,
+		"type.googleapis.com/envoy.extensions.load_balancing_policies.wrr_locality.v3.WrrLocality":                                      convertWRRLocalityProtoToServiceConfig,
+		"type.googleapis.com/envoy.extensions.load_balancing_policies.client_side_weighted_round_robin.v3.ClientSideWeightedRoundRobin": convertWeightedRoundRobinProtoToServiceConfig,
+		"type.googleapis.com/xds.type.v3.TypedStruct":                                                                                   convertV3TypedStructToServiceConfig,
+		"type.googleapis.com/udpa.type.v1.TypedStruct":                                                                                  convertV1TypedStructToServiceConfig,
+	}
+}
```
Comment (dfawley) [inline xds/internal/xdslbregistry/converter.go:61]: Combine, delete init:

```go
var m = map[string]converter{
...
}
```

**Review by dfawley — APPROVED** [review]

File: xds/internal/xdslbregistry/converter_test.go (hunk 18 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter_test.go]: I deleted this file and moved to e2e test.

File: test/xds/xds_client_custom_lb_test.go [hunk 23]
```diff
@@ -22,20 +22,25 @@ import (
 	"context"
 	"fmt"
 	"testing"
+	"time"
 
 	v3xdsxdstypepb "github.com/cncf/xds/go/xds/type/v3"
```
[the remark this answers is not in the document]
Comment (zasweq) [reply test/xds/xds_client_custom_lb_test.go:27]: Done.

File: xds/internal/xdslbregistry/converter.go (hunk 21 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:43]: Moved above.

File: xds/internal/xdslbregistry/converter.go (hunk 22 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/converter.go:61]: As per team thread, couldn't declare inline due to initialization cycle.

File: xds/internal/xdslbregistry/tests/converter_test.go (hunk 20 above)
Comment (zasweq) [reply xds/internal/xdslbregistry/tests/converter_test.go:20]: Moved as per convention.
